File: ruby/version2_traduccion.rb
# http://localhost:4568/?n=10
# ruby version2_traduccion.rb

require 'sinatra'
require 'savon'
require 'httparty'
require 'uri'

set :port, 4568

get '/' do
  n = params['n']
  halt 400, 'Proporciona un número: http://localhost:4568/?n=10' if n.nil?

  # Llamada SOAP
  client = Savon.client(wsdl: 'https://www.dataaccess.com/webservicesserver/NumberConversion.wso?WSDL')
  response = client.call(:number_to_words, message: { ubi_num: n.to_i })
  en_ingles = response.body[:number_to_words_response][:number_to_words_result]

  # Traducción con MyMemory API
  url = "https://api.mymemory.translated.net/get?q=#{URI.encode_www_form_component(en_ingles)}&langpair=en|es"
  trad_response = HTTParty.get(url)
  traducido = trad_response['responseData']['translatedText']

  content_type 'text/plain; charset=utf-8'
  traducido
end

File: perl/version2_traduccion.pl
# http://localhost:7071/?n=10
# perl version2_traduccion.pl

use strict;
use warnings;
use SOAP::Lite;
use LWP::UserAgent;
use HTTP::Daemon;
use HTTP::Status;
use URI;
use URI::Escape;
use JSON;

my $servidor = HTTP::Daemon->new(
    LocalAddr => 'localhost',
    LocalPort => 7071,
) or die "No se pudo iniciar el servidor: $!";

print "Servidor en http://localhost:7071/?n=10\n";

while (my $conexion = $servidor->accept()) {
    while (my $peticion = $conexion->get_request()) {
        my $uri = URI->new($peticion->uri);
        my %params = $uri->query_form;
        my $n = $params{n} // '0';

        # Llamada SOAP
        my $cliente = SOAP::Lite->service(
            'https://www.dataaccess.com/webservicesserver/NumberConversion.wso?WSDL'
        );
        my $en_ingles = $cliente->NumberToWords($n);

        # Traducción con MyMemory API
        my $ua = LWP::UserAgent->new();
        my $url = "https://api.mymemory.translated.net/get?q=" .
                  uri_escape($en_ingles) . "&langpair=en|es";

        my $trad_resp = $ua->get($url);
        my $datos = decode_json($trad_resp->content);
        my $traducido = $datos->{responseData}{translatedText};

        my $respuesta = HTTP::Response->new(200);
        $respuesta->header('Content-Type' => 'text/plain; charset=utf-8');
        $respuesta->content($traducido);
        $conexion->send_response($respuesta);
    }
    $conexion->close();
}